Apps/Src/Tasks.c

```diff
@@ -45,26 +46,116 @@ CPU_STK DebugDump_Stk[TASK_DEBUG_DUMP_STACK_SIZE];
 CPU_STK CommandLine_Stk[TASK_COMMAND_LINE_STACK_SIZE];
 
 /**
- * Semaphores
+ * Global Variables
  */
-OS_SEM FaultState_Sem4;
+os_error_loc_t OSErrLocBitmap = OS_NONE_LOC; // Store the location of the current error
+
+// Variables to store error codes, stored and cleared in task error assert functions
+error_code_t Error_ReadCarCAN = /*READCARCAN_ERR_NONE*/ 0; // TODO: change this back to the error 
+error_code_t Error_ReadTritium = T_NONE;  // Initialized to no error
+error_code_t Error_UpdateDisplay = UPDATEDISPLAY_ERR_NONE;
+
+extern const pinInfo_t PININFO_LUT[]; // For GPIO writes. Externed from Minions Driver C file.
 
 /**
- * Global Variables
- */
-fault_bitmap_t FaultBitmap = FAULT_NONE;
-os_error_loc_t OSErrLocBitmap = OS_NONE_LOC;
+ * Error assertion-related functions
+*/
 
-void _assertOSError(uint16_t OS_err_loc, OS_ERR err)
+void _assertOSError(os_error_loc_t OS_err_loc, OS_ERR err)
 {
     if (err != OS_ERR_NONE)
     {
-        FaultBitmap |= FAULT_OS;
-        OSErrLocBitmap |= OS_err_loc;
+        EmergencyContactorOpen(); // Turn off contactors and turn on the brakelight to indicate an emergency
+        Display_Error(OS_err_loc, err); // Display the location and error code
+        while(1){;} //nonrecoverable
+
+    }
+}
+
+/**
+ * @brief Assert a task error by locking the scheduler (if necessary), displaying a fault screen,
+ * and jumping to the error's specified callback function. 
+ * Called by task-specific error-assertion functions that are also responsible for setting the error variable.
+ * @param errorLoc the task from which the error originated. Note: should be taken out when last task pointer is integrated
+ * @param errorCode the enum for the specific error that happened
+ * @param errorCallback a callback function to a handler for that specific error, 
+ * @param lockSched whether or not to lock the scheduler to ensure the error is handled immediately. Only applicable for recoverable errors- nonrecoverable errors will always lock
+ * @param nonrecoverable whether or not to kill the motor, display the fault screen, and enter an infinite while loop
+*/
+void throwTaskError(os_error_loc_t errorLoc, error_code_t errorCode, callback_t errorCallback, error_scheduler_lock_opt_t lockSched, error_recov_opt_t nonrecoverable) {
+    OS_ERR err;
+
+    if (errorCode == 0) { // Exit if there is no error
+        return;
+    }
+
+    if (lockSched == OPT_LOCK_SCHED || nonrecoverable == OPT_NONRECOV) { // Prevent other tasks from interrupting the handling of important (includes all nonrecoverable) errors
```

Comment:
Even if the scheduler is locked, interrupts are still enabled so it could be interrupted by something else (which is something that happened in the BPS). Also why would something be nonrecoverable and not have OPT_LOCK_SHED? Shouldn't this just happen if OP_LOCK_SCHED is true? If they are not mutually exclusive, then shouldn't you just get rid of OS_LOCK_SCHED and only lock when the error is non-recoverable?

Reply:
I'm sort of seeing that this entire thing would run only if the error code was OPT_NONRECOV if you did the above, which makes me think that this is just a fancier assertOSerror (where it runs only if the error is an error). So you could maybe model this function off of assertOSerror.

Reply:
Yeah, the OPT_LOCK_SCHED and nonrecoverable thing is kind of weird. It's because we wanted some errors to lock the scheduler without being nonrecoverable, but we also wanted to make everything nonrecoverable also lock the scheduler without choice. Sort of a squares and rectangles situation, I guess. We'll work on this in M3 🫡